### PR TITLE
Jetpack Block PHP Fixture Tests: Add test for the server rendered Business Hours block

### PR DESCRIPTION
**Pull request by andrewserong**
Related to: 251-gh-Automattic/view-design

With the recent work to add in [block test fixtures](https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/extensions/shared/test/block-fixtures.md) we noticed a gap in the testing for blocks that use server rendering. In addition to testing that the block's JS save method generates consistent markup, we also need to ensure that the serialized markup continues to result in correctly rendered server-side markup, when the block is rendered in PHP.

To explore one approach for testing fixtures for server-rendered blocks, this PR proposes adding in a test for the Business Hours block, since it has a fairly straightforward server rendering approach in that it doesn't call to any outside endpoints to generate its markup.

The approach I've explored involves:

The Business Hours block PHP test generates corresponding `*.server-rendered.html` fixtures for each of the block's `*.serialized.html` fixtures if they're not present. These generated fixtures are intended to be committed to the repo.

When matching generated fixtures exist, the PHP test checks the server-rendered `render_block` results against this fixture.

While working on writing or updating tests, a developer can run `yarn docker:phpunit --testsuite=extensions` to either test the existing fixtures, or if fixtures are missing, generate them.

#### Changes proposed in this Pull Request:

* For the Business Hours block, load all serialized fixtures matching the pattern `*.serialized.html`
* For each of these fixtures, check to see if a matching `*.server-rendered.html` fixture exists.
* If not, create one by passing the `*.serialized.html` fixture to `render_blocks` and save to disk. Then error out of the test to flag to the developer that the fixture file has been created and will need to be committed.
* If a fixture file does exist, then check to see that the fixture file's markup matches what's returned by calling `render_blocks` with the serialized markup for the block.

#### Jetpack product discussion

If this approach seems viable, then I can update the block fixtures documentation with a link to this example in a separate PR, and we can look at using this approach for other server-rendered blocks, too.

#### Testing instructions:

* Ensure the PHP tests are passing in CI

##### Testing locally

1. Check out this change locally, and following the instructions in this repo, set up your local Docker development environment and run the following from the root directory of this repo:

```
yarn docker:phpunit --testsuite=extensions
```

2. Delete the generated server-rendered fixtures included in this PR, and run the above command again. You should see that the test throws an error, and reports back the fixtures that have been generated:

image

3. Run the command again, and the tests should pass
4. Manually make a change to one of the server-rendered fixtures files and re-run the testing command. It should fail, complaining that the strings don't match. But, it should also tell you _which_ fixture it failed on with a (hopefully) friendly message than just the strings not matching:

image

Also, a question for you! Does this approach to server rendered fixtures sound reasonable? Is there anything missing, or is there another approach that you think would work better instead?

### CODE REVIEW

File: projects/plugins/jetpack/tests/php/extensions/blocks/test-business-hours.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+	 * This test iterates over the block's serialized fixtures, and tests that the generated
+	 * markup matches a fixture for the server rendered markup for the block.
+	 *
+	 * If no server rendered fixture can be found, then one is created.
+	 */
+	public function test_server_side_rendering_based_on_serialized_fixtures() {
+		// phpcs:disable WordPress.WP.AlternativeFunctions
+		$fixtures_path = 'extensions/blocks/business-hours/test/fixtures/';
+		$file_pattern  = '*.serialized.html';
+		$files         = glob( JETPACK__PLUGIN_DIR . $fixtures_path . $file_pattern );
```
Comment (ramonjd) [inline projects/plugins/jetpack/tests/php/extensions/blocks/test-business-hours.php:72]: Not for this PR, but if we get the 🟢 for these tests this and some of the surrounding code blocks would be handy as a shared util or even custom assertion method.

**Comment by ramonjd** [comment]
I love this. Thanks for working on it @andrewserong 

The tests pass (and fail) as expected for me. 

I also picked a random block (Calendly) that had a `render_callback` and copied the test over. I think Calendly requires some extra mocking or something but it works nearly out of the box!

File: projects/plugins/jetpack/tests/php/extensions/blocks/test-business-hours.php (hunk 1 above)
Comment (andrewserong) [reply projects/plugins/jetpack/tests/php/extensions/blocks/test-business-hours.php:72]: Thanks, I was thinking of that, too. I wasn't quite sure how much each block might vary in their implementation, so thought refactoring to utils could be a good thing to do if and when this one's approved and we work on the second (or third) test.

**Review by ramonjd — APPROVED** [review]
A great first step. I think it's good to go, seeing as it focusses initially on one block. 

Thank you! 🥇

**Comment by andrewserong** [comment]
Thanks for testing @ramonjd!

> One question: does it handle deprecations, or do we even need to?

That's a good question! For _this_ block, we don't (yet) need to worry about it because we don't have any deprecations. The goal currently is to compare the serialized markup against the server rendered markup. For a block with deprecations, we could get it to compare against say `jetpack__calendly__deprecated-1.html` as well as `jetpack__calendly__deprecated-1.serialized.html`. But I think you're right — the server rendering needs to be able to deal with markup that has been deprecated but not yet migrated to the expected current markup (so that existing blocks don't break on a user's site). I think I'd be inclined to deal with deprecations in a follow-up when we attempt to apply the same approach as this PR to another block with deprecations.

But also happy to look into it here, if you (or anyone else, too) think it's better to come up with an approach from the outset!

**Comment by ramonjd** [comment]
> But also happy to look into it here, if you (or anyone else, too) think it's better to come up with an approach from the outset!

I only mentioned it because I tried it out with Calendly, which has a deprecation. I suppose it would make more sense to tackle that little number later as we need it 😆

**Review by jeherve — APPROVED** [review]

**Comment by andrewserong** [comment]
r225522-wpcom

**Pull request merged**